src/vivarium/framework/components/manager.py

```diff
@@ -26,6 +26,40 @@ class ComponentConfigError(VivariumError):
     pass
 
 
+class ComponentList(list):
+    """Container for Vivarium components that enforces uniqueness according to name."""
+
+    def _check_conditions(self, component):
+        if not hasattr(component, "name"):
+            raise ComponentConfigError(f"Attempting to add a component {component} with no name. Vivarium "
+                                       "requires each component in a simulation have a name attribute with"
+                                       "a unique name.")
+        if component in self:
+            raise ComponentConfigError(f"Attempting to append component with duplicate name: {component}")
+
+    def append(self, component) -> None:
+        self._check_conditions(component)
+        super().append(component)
+
+    def extend(self, components):
+        for c in components:
+            self.append(c)
+
+    def insert(self, i, component):
+        self._check_conditions(component)
+        super().insert(i, component)
+
+    def __setitem__(self, key, component):
```

Comment:
I'm not sure about using the same check_conditions method for setitem. Won't this throw the 'attempting to append' error if I try to replace a component with the same name? e.g., if index 1 is a component with name 'NAME' and I try to set index 1 to a new component also with name 'NAME', I'll get an error. Or is that not a use case we think will come up?